packages/transformer/src/IModelExporter.ts

```diff
@@ -942,82 +927,167 @@ export class ChangedInstanceIds {
+    // if changeType is a delete and we already have the id in the inserts then we can remove the id from the inserts.
+    // if changeType is a delete and we already have the id in the updates then we can remove the id from the updates AND add it to the deletes.
+    // if changeType is an insert and we already have the id in the deletes then we can remove the id from the deletes AND add it to the inserts.
+    if (changeType === "Inserted") {
+      changedInstanceOps.insertIds.add(id);
+      changedInstanceOps.deleteIds.delete(id);
+    } else if (changeType === "Updated") {
+      if (!changedInstanceOps.insertIds.has(id))
+        changedInstanceOps.updateIds.add(id);
+    } else if (changeType === "Deleted") {
+      // If we've inserted the entity at some point already and now we're seeing a delete. We can simply remove the entity from our inserted ids without adding it to deletedIds.
+      if (changedInstanceOps.insertIds.has(id))
+        changedInstanceOps.insertIds.delete(id);
+      else {
+        changedInstanceOps.updateIds.delete(id);
+        changedInstanceOps.deleteIds.add(id);
+      }
+    }
+   }
 
   /**
    * Initializes a new ChangedInstanceIds object with information taken from a range of changesets.
    */
-  public static async initialize(opts: ChangedInstanceIdsInitOptions): Promise<ChangedInstanceIds>;
-  /**
-   * Initializes a new ChangedInstanceIds object with information taken from a range of changesets.
-   * @deprecated in 0.1.x. Pass a [[ChangedInstanceIdsInitOptions]] object instead of a changeset id
-   * @param accessToken Access token.
-   * @param iModel IModel briefcase whose changesets will be queried.
-   * @param firstChangesetId Changeset id.
-   * @note Modified element information will be taken from a range of changesets. First changeset in a range will be the 'firstChangesetId', the last will be whichever changeset the 'iModel' briefcase is currently opened on.
-   */
-  public static async initialize(
-    accessTokenOrOpts: AccessToken | ChangedInstanceIdsInitOptions | undefined,
-    iModel?: BriefcaseDb,
-    startChangesetId?: string,
-  ): Promise<ChangedInstanceIds> {
-    const opts: ChangedInstanceIdsInitOptions
-      = typeof accessTokenOrOpts === "object"
-      ? accessTokenOrOpts
-      : {
-        accessToken: accessTokenOrOpts,
-        iModel: iModel!,
-        startChangeset: { id: startChangesetId! },
-      };
-
-    assertHasChangeDataOptions(opts);
-
+  public static async initialize(opts: ChangedInstanceIdsInitOptions): Promise<ChangedInstanceIds> {
     const iModelId = opts.iModel.iModelId;
     const accessToken = opts.accessToken;
 
-    const changesetRanges = opts.changesetRanges
-      ?? [[
-        // we know startChangeset is defined because of the assert above
-        opts.startChangeset!.index
-          ?? (await IModelHost.hubAccess.queryChangeset({
+    let csFileProps: ChangesetFileProps[];
```

Comment:
I gave it a shot, let me know what you think. Also relevant to my changes is this comment you made:

"honestly we might want to refactor this into a private initializeFromChangeOptions function, which takes the options, determines which were passed, and returns the ChangedInstanceIds object if necessary after creating. Then hasChangeData can just be a check as to whether that returned undefined." 

Instead of adding the private initFromChangeOptions, I did below instead:
I made it so that ChangedInstanceIds.initialize can return undefined (which might not be allowed? maybe it is since we're going to 1.x?)
and I just check all cases in there, i.e. if opts.changedinstanceids is defined return early. I also get to check the case where there are none of the 4 (changesetranges, startchangeset, changedinstanceids, csFileProps) and return undefined.

Reply:
your redo LGTM

Reply:
Changing to throw instead of return undefined causes lots of tests to fail. Its technically valid to pass no argument isn't it(aka none of the 4: startChangeset, changesetRanges, csFileProps, changedinstanceIds)?. Atleast thats how the type is setup right now. Also some cases like exportAll calls this.initialize() with {}

```
 /** Export all entity instance types from the source iModel.
   * @note [[exportSchemas]] must be called separately.
   */
  public async exportAll(): Promise<void> {
    await this.initialize({});
```

Reply:
iirc there was a default behavior for backwards compat. I suppose we can keep it as optional and keep that behavior, I don't remember it being bad. I'll investigate that just in case tomorrow